keras_nlp/tokenizers/byte_pair_tokenizer.py

```diff
@@ -0,0 +1,460 @@
+        return result
+
+    def _remove_empty_strings(self, tensor):
+        """Remove empty strings in a tensor"""
+        return self._remove_strings(tensor, "")
+
+    def _remove_whitespace_placeholder(self, tensor):
+        return self._remove_strings(tensor, "६")
+
+    @tf.function
+    def _byte_pair_merge_loop_body(self, words, mask):
+        """Iterative merging process for byte pair encoding algorithm.
+
+        The end condition is either the word has been fully merged (list has
+        only one byte string), or it can no longer perform a merge.
+        """
+        # Get all word pairs.
+        first, second = words[:, :-1], words[:, 1:]
+
+        # Mask empty.
+        non_empty_mask = second.nested_row_lengths()[0] != 0
+        mask = tf.logical_and(mask, non_empty_mask)
+        if tf.math.count_nonzero(mask) == 0:
+            return [words, mask]
+        non_empty_idxs = tf.boolean_mask(tf.range(tf.shape(mask)[0]), mask)
+        tmp_first = tf.ragged.boolean_mask(first, mask)
+        tmp_second = tf.ragged.boolean_mask(second, mask)
+
+        # Get byte pair ranking in merge rules.
+        pairs = tf.strings.join([tmp_first, tmp_second], separator=" ")
+        pair_rank = self.bpe_ranks.lookup(pairs)
+
+        # Get BPE pair ranks.
+        min_pair_rank = tf.reduce_min(pair_rank, axis=1)
+        not_found_mask = min_pair_rank != self.max_bpe_rank
+
+        # Tokens cannot be further merged are marked as finished.
+        mask = tf.tensor_scatter_nd_update(
+            mask, tf.expand_dims(non_empty_idxs, axis=1), not_found_mask
+        )
+        if tf.math.count_nonzero(mask) == 0:
+            return [words, mask]
+
+        masked_pair_rank = tf.ragged.boolean_mask(pair_rank, not_found_mask)
+        min_pair_rank_idx = tf.math.argmin(
+            masked_pair_rank.to_tensor(self.max_bpe_rank), axis=1
+        )
+
+        # Get words and pairs to process.
+        unfinished_words = tf.ragged.boolean_mask(words, mask)
+
+        pair_left = tf.gather(unfinished_words, min_pair_rank_idx, batch_dims=1)
+        pair_right = tf.gather(
+            unfinished_words, min_pair_rank_idx + 1, batch_dims=1
+        )
+
+        merged_pairs = tf.strings.join([pair_left, pair_right])
+        empty_strs = tf.fill(tf.shape(merged_pairs), "")
+
+        unfinished_word_indices = tf.cast(
+            tf.boolean_mask(tf.range(tf.shape(mask)[0]), mask), dtype=tf.int64
+        )
+        merged_pair_indices = tf.concat(
+            [
+                unfinished_word_indices[:, tf.newaxis],
+                min_pair_rank_idx[:, tf.newaxis],
+            ],
+            axis=1,
+        )
+        empty_string_indices = tf.concat(
+            [
+                unfinished_word_indices[:, tf.newaxis],
+                min_pair_rank_idx[:, tf.newaxis] + 1,
+            ],
+            axis=1,
+        )
+
+        tensor_words = words.to_tensor(default_value="")
+        tensor_words = tf.tensor_scatter_nd_update(
```

Comment:
could we combine these into a single scatter? that seems like it might buy us some efficiency.

is there a reason they need to be sequential?

Reply:
for some reason the auto conversion failed:
```
E             TypeError: Failed to convert elements of tf.RaggedTensor(values=Tensor("words:0", shape=(None,), dtype=string), row_splits=Tensor("words_1:0", shape=(5,), dtype=int64)) to Tensor. Consider casting elements to a supported type. See https://www.tensorflow.org/api_docs/python/tf/dtypes for supported TF dtypes.
```